### PR TITLE
Broken build under MSVC

### PR DESCRIPTION
**Pull request by bkentel**
This is one of a few bugs left in MSVC's c++11 conformance. Luckily
there is usually an easy work around such as this.

### CODE REVIEW

File: src/ui.h [hunk 1]
```diff
@@ -38,7 +38,7 @@ struct uimenu_entry {
     std::string txt;      // what it says on the tin
     nc_color hotkey_color;
     nc_color text_color;
-    mvwzstr extratxt = { 0, c_unset, "", 0};
+    mvwzstr extratxt = mvwzstr { 0, c_unset, "", 0};
```
Comment (narc0tiq) [inline src/ui.h]: If you really want to maintain DRY, that may also work as `auto extratxt = mvwzstr { initializer_stuff }`

**Comment by hbnix** [comment]
Oops, that was me, sorry :/ Will keep this in mind.

File: src/ui.h (hunk 1 above)
Comment (bkentel) [inline src/ui.h]: Don't think `auto` works for [field declarations](http://stackoverflow.com/questions/11302981/c11-declaring-non-static-data-members-as-auto) unfortunately(?); this is unfortunately a verbose workaround needed until `initializer lists` are 100% supported in MSVC.

The other big hole at the moment is full `constexpr` support (mostly use inside and on class/struct declarations).

**Comment by bkentel** [comment]
No problem :smile:  can't expect everyone to know what MS [hasn't got working yet](http://blogs.msdn.com/b/vcblog/archive/2014/11/17/c-11-14-17-features-in-vs-2015-preview.aspx).

**Comment by KA101** [comment]
||=== Build: Debug in Cataclysm (compiler: GNU GCC Compiler) ===|
C:\Documents and Settings\---\My Documents\GitHub\Cataclysm-DDA\src\ui.h||In constructor 'uimenu_entry::uimenu_entry(std::string, int)':|
C:\Documents and Settings\---\My Documents\GitHub\Cataclysm-DDA\src\ui.h|47|internal compiler error: in gimple_add_tmp_var, at gimplify.c:716|
||=== Build failed: 1 error(s), 0 warning(s) (5 minute(s), 28 second(s)) ===|


Motion to reinstate non-support of MSVC.

**Comment by bkentel** [comment]
I'll try a different workaround later. Which version of GCC are you using by the way?

**Comment by hbnix** [comment]
Hm, for me the fix compiles, using mingw-w64 gcc-4.8.1

**Comment by KA101** [comment]
mingw32, default C::B.  I know that -w64 can handle it and should have a working setup--though that then changed my project file, which then wants to backflow up to master.

**Comment by bkentel** [comment]
@KA101 Right, mingw32, but what version of gcc? `internal compiler error` always indicates a compiler bug. For that matter does the project have a target for "oldest version of compiler XYZ we intend to support"?

**Comment by kevingranade** [comment]
Yes, rough notes are in doc/COMPILER_SUPPORT.md

**Comment by KA101** [comment]
According to the readme in mine, 4.6-4.7.

**Comment by bkentel** [comment]
@KA101 can you do a `gcc --version`?

**Comment by bkentel** [comment]
@KA101 [looks like 4.8.1](http://sourceforge.net/projects/mingw/files/MinGW/Base/gcc/Version4/) is the newest supported version of gcc for mingw32 -- why not upgrade?

File: src/ui.h (hunk 1 above)
Comment (kevingranade) [inline src/ui.h]: As this is evidently fragile depending on the compiler, how about you just put it in the constructors?

**Comment by KA101** [comment]
Chiefly because mucking about with an upgrade is the last thing I want to start doing at this time of night.  Might look into it tomorrow or Monday, but getting -w64 running was enough of a PitA IMO.

**Comment by kevingranade** [comment]
If we have a choice between supporting slightly older compiler versions and making just one or two places in the code less tidy than they could be, we're definitely going with maintaining support.

File: src/ui.h (hunk 1 above)
Comment (bkentel) [inline src/ui.h]: That's also a way of fixing it.

**Comment by bkentel** [comment]
Fair enough, but it's not until 4.8 that GCC becomes c++11 feature complete.
https://gcc.gnu.org/projects/cxx0x.html

I'll change the fix to use a standard constructor if gcc 4.8+ is deemed to be too restrictive.

**Comment by kevingranade** [comment]
If complete c++11 support is the baseline, we would be rejecting your VS
patches, it's never going to be c++11 complete.  We're supporting the most
compilers we can that support the subset of c++11 we use.  We make minimal
changes to maintain as much compiler support as is reasonable.
E.g. if someone has a strong case for using a lot of constexpr, VS is going
to go unsupported again.

**Comment by bkentel** [comment]
Just playing devils advocate here, but here the feature exists in a widely-available GCC, with MSVC no such option exists (as yet).

The constructor workaround is the best option here, I agree.

**Comment by kevingranade** [comment]
'A widely available gcc' doesn't do a thing if it's not the version that
comes with the platform.  What we want is as low a barrier to entry as
possible for the majority of contributors, very many of which use C::B and
mingw.  AFAICT we've only had maybe 5 VS contributors as opposed to 100+
C::B/mingw contributors.

**Comment by bkentel** [comment]
But this *is* a version that comes with the platform. Mainline mingw32 uses 4.8.1 and the experimental version 4.9.x (even the DDA wiki points to these).

I'm not saying you're wrong or off base about trying to be as accessible as possible, but at least as far as mingw is concerned the point is moot.

**Comment by KA101** [comment]
FWIW, Code::Blocks' latest release (I'm using it) ships with 4.7.1 as the endorsed mingw/GCC version.
http://www.codeblocks.org/downloads/26
Specifically: "While v4.7.1 is rock-solid (we use it to compile C::B), v4.8.1 is provided for convenience, there are some known bugs with this version related to the compilation of Code::Blocks itself.  IF UNSURE, USE "codeblocks-13.12mingw-setup.exe"! "

New folks can safely be presumed to be Unsure, I'd imagine.

**Comment by bkentel** [comment]
Didn't realize C::B had a link to mingw; the link on the DDA wiki leads to the 4.8.1 version which is also the default version one gets via the mingw site.

**Comment by bkentel** [comment]
Added a different fix.

**Comment by i2amroy** [comment]
Another note, but Cygwin doesn't totally support 4.8.1 without rather significant work on the user part ATM either as far as I know, so limiting things to 4.8.1 would cause some rather significant problems there as well.

**Comment by bkentel** [comment]
Not sure exactly what version of GCC this bug is an issue for, but I added a fix that should be ok if anyone using an older version of GCC wants to test it out.

**Pull request merged**